### PR TITLE
Feat/authentication/view model : View Model for Supabase Authentication

### PR DESCRIPTION
**Pull request by coaguila**
# View Model for Supabase Authentication
## Description
This PR introduces the View Model for the Supabase Authentication. It adds multiple new utilities classes. For now it only accepts user account creation through email. Add Also a rule for the `MainCoroutine` to be able to run test on functions that have the `viewModelScope`. It closes issue #11. 
## Changes
`AuthModelSupabase` function `isLoggedIn` refactored for readibility. `MainCoroutineRule.kt` created to test functions that have `viewModelScope`.
#### Added
- `AuthViewModel.kt` in `app.src.main,java,com.android.periodpals.model.auth`
- `AuthViewModelTest.kt` in `app.src.test,java,com.android.periodpals.model.auth`
- `MainCoroutineRule.kt` in `app.src.test,java,com.android.periodpals`
- `SupabaseClient.kt` in `app.src.test,java,com.android.periodpals.model`
- `SharedPreferenceHelper.kt` in `app/src/main/java/com/android/periodpals/utils`
- `User.kt` in `app/src/main/java/com/android/periodpals/model/user`
- `UserAuthState.kt` in `app/src/main/java/com/android/periodpals/model/user`
#### Modified
- `app/build.gradle.kts`
- `AuthModelSupabase.kt` in `app.src.main,java,com.android.periodpals.model.auth`
- `AuthModelSupabaseTest.kt` in `app.src.test,java,com.android.periodpals.model.auth`
## Dependencies Added
- `org.jetbrains.kotlinx:kotlinx-coroutines-test:1.6.1` for `MainCoroutineRule.kt`
## Testing
- Added testing for `AuthViewModel.kt`. It tests wether it correctly handled successful and erroneous calls to the authentication model.
- Modified testing for `AuthModel.kt` to adapt correctly to the changes to the class.

### CODE REVIEW

File: app/src/main/java/com/android/periodpals/utils/SharedPreferenceHelper.kt
Comment (francelu) [inline app/src/main/java/com/android/periodpals/utils/SharedPreferenceHelper.kt:1]: This class is never used. Please define its utility.

File: app/src/test/java/com/android/periodpals/MainCoroutineRule.kt
Comment (francelu) [inline app/src/test/java/com/android/periodpals/MainCoroutineRule.kt:1]: LGTM. Please document this.

**Review by francelu — CHANGES_REQUESTED** [review]
LGTM.  
Requires minor changes for PR approval, c.f. code / file comments.

File: app/src/main/java/com/android/periodpals/utils/SharedPreferenceHelper.kt
Comment (coaguila) [reply app/src/main/java/com/android/periodpals/utils/SharedPreferenceHelper.kt:1]: I agree will delete. 👍🏽

File: app/src/main/java/com/android/periodpals/model/user/User.kt
[the remark this answers is not in the document]
Comment (coaguila) [reply app/src/main/java/com/android/periodpals/model/user/User.kt:1]: I agree will delete. 👍🏽

File: app/src/test/java/com/android/periodpals/MainCoroutineRule.kt
Comment (coaguila) [reply app/src/test/java/com/android/periodpals/MainCoroutineRule.kt:1]: done

**Review by Harrish92 — APPROVED** [review]
LGTM.

**Review by francelu — APPROVED** [review]
LGTM. Thanks !

**Pull request merged**